crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -70,22 +69,30 @@ pub(crate) async fn fast_sync<Backend, Block, AS, Client, PG, NR>(
             import_queue_service.as_mut(),
             &network_request,
             &sync_service,
-        )
-        .await;
+        );
 
-        match fast_sync_result {
+        match fast_sync_fut.await {
             Ok(()) => {
                 debug!("Fast sync finished successfully");
             }
-            Err(err) => {
-                error!("Fast sync failed: {err}");
+            Err(error) => {
+                error!(%error, "Fast sync failed");
             }
         }
+
+        pause_sync.store(false, Ordering::Release);
     } else {
-        debug!("Fast sync detected existing finalized hash.");
+        debug!("Fast sync can only work with genesis state, skipping");
     }
 
-    pause_sync.store(false, Ordering::Release);
+    // Switch back to full sync mode
+    let info = client.info();
+    sync_service
```

Comment:
This change does seem problematic: we unpause regular sync before the gap sync removal, and that can cause problems. The symtomps of which I got during the restart test:
```
2024-05-29T11:30:38.775050Z  INFO Consensus: substrate: ⏩ Block history, #0 (1 peers), best: #1724666 (0xc468…b82c), finalized #0 (0x0c12…1c34), ⬇ 7.4kiB/s ⬆ 5.8kiB/s    
2024-05-29T11:30:39.741072Z  INFO Consensus: substrate: ✨ Imported #1724667 (0x1a41…ccac)    
2024-05-29T11:30:43.779729Z  INFO Consensus: substrate: ⏩ Block history, #0 (2 peers), best: #1724667 (0x1a41…ccac), finalized #0 (0x0c12…1c34), ⬇ 8.7kiB/s ⬆ 8.2kiB/s    
2024-05-29T11:30:44.308137Z  INFO Consensus: substrate: ✨ Imported #1724668 (0x7d79…10cf)    
2024-05-29T11:30:48.780323Z  INFO Consensus: substrate: ⏩ Block history, #0 (0 peers), best: #1724668 (0x7d79…10cf), finalized #0 (0x0c12…1c34), ⬇ 10.2kiB/s ⬆ 9.4kiB/s    
2024-05-29T11:30:50.431849Z  INFO Consensus: substrate: ✨ Imported #1724669 (0xf92b…e0b5)    
2024-05-29T11:30:50.440795Z  INFO Consensus: substrate: ✨ Imported #1724670 (0xc9f2…a47a)    
2024-05-29T11:30:53.780437Z  INFO Consensus: substrate: ⏩ Block history, #0 (0 peers), best: #1724670 (0xc9f2…a47a), finalized #0 (0x0c12…1c34), ⬇ 9.1kiB/s ⬆ 8.3kiB/s    
2024-05-29T11:30:56.452659Z  INFO Consensus: substrate: ✨ Imported #1724671 (0xb8f0…e4e1)    
2
```

Reply:
The old way was problematic, I fixed it, but forgot to fix incorrect condition check in `subspace-service` that was similar to above. Pushed commit with a fix.

> we unpause regular sync before the gap sync removal

We are not doing that. One clearing happened before this sync code is even triggered in `subspace-service` and another in sync function below, which is also before.

Reply:
I see it now, makes sense.